src/backtracking/rat_in_maze.rs

```diff
@@ -0,0 +1,267 @@
+//! This module contains the implementation of the Rat in Maze problem.
+//!
+//! The Rat in Maze problem is a classic algorithmic problem where the
+//! objective is to find a path from the starting position to the exit
+//! position in a maze.
+
+/// Finds a path through the maze starting from the specified position.
+///
+/// # Arguments
+///
+/// * `maze` - The maze represented as a vector of vectors where each
+/// inner vector represents a row in the maze grid.
+/// * `start_x` - The x-coordinate of the starting position.
+/// * `start_y` - The y-coordinate of the starting position.
+///
+/// # Returns
+///
+/// A solution matrix if a path is found or `None` if not found.
+/// During the maze traversal process, the rat explores different paths,
+/// marking its progress in the solution matrix. The goal is to find a path
+/// from the starting position to the exit (end) of the maze. Once the rat
+/// successfully reaches the exit, the solution matrix will contain the path
+/// taken by the rat, enabling us to reconstruct the solution and visualize the
+/// rat's journey through the maze.
+pub fn find_path_in_maze(
+    maze: &[Vec<bool>],
+    start_x: usize,
+    start_y: usize,
+) -> Option<Vec<Vec<bool>>> {
+    let mut maze_instance = Maze::new(maze.to_owned());
+    maze_instance.find_path(start_x, start_y)
+}
+
+/// Represents a maze.
+struct Maze {
+    maze: Vec<Vec<bool>>,
+}
+
+impl Maze {
+    /// Represents possible moves in the maze.
+    const MOVES: [(isize, isize); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
+
+    /// Constructs a new Maze instance.
+    /// # Arguments
+    ///
+    /// * `maze` - The maze represented as a vector of vectors where each
+    /// inner vector represents a row in the maze grid.
+    ///
+    /// # Returns
+    ///
+    /// A new Maze instance.
+    fn new(maze: Vec<Vec<bool>>) -> Self {
+        Maze { maze }
+    }
+
+    /// Returns the width of the maze.
+    ///
+    /// # Returns
+    ///
+    /// The width of the maze.
+    fn width(&self) -> usize {
+        self.maze.iter().map(|row| row.len()).max().unwrap_or(0)
```

Comment:
This is quite complex. Why don't you just return some error if the input is not a rectangle?

Or, since this project [depends on `nalgebra`](https://github.com/TheAlgorithms/Rust/blob/864ef42a427cb1f841d85c60b96c01c63154e54c/Cargo.toml#L13), you could use some matrix type provided there.

I prefer the first approach.

Reply:
What do you mean by returning some error if the input is not a rectangle? Currently, the implementation handles the maze in both rectangle and non-rectangle representation. This function returns the width of the maze, which is equal to the largest row in the maze.

Reply:
The signature of `find_path_in_maze` could look like:
```rust
pub fn find_path_in_maze(
    maze: &[Vec<bool>],
    start_x: usize,
    start_y: usize,
) -> Result<Option<Vec<Vec<bool>>>, SomeInvalidInputError>
```

Reply:
Could you review the update, @vil02. I have adjusted the implementation of the function `find_path_in_maze` to handle exception cases: empty maze, start out-of-bound on the maze, and non-rectangle maze. I also added some edge tests to cover various scenarios of maze pathfinding.